### PR TITLE
Fixed #429 - Memory issue on system page.

### LINKED ISSUE #429: Memory SVG on System page incorrect for certain Unix and docker systems
Server side

When fixing #378 I already noticed weird scaling and missing child property `mem`. Now the error is gone, but the calculation for non `memRss` calculations is off (I didnt change anything here).
Needs to be determined where the problem lies and rectified for that situation. I expect somehow `mem` is a different unit than kB (f.e. MB's or bytes).

### PR DESCRIPTION
**Pull request by NotExpectedYet**
Confirm working on: 

 - [x] Ubuntu
 - [x] Docker Alpine
 - [x] Docker Monolithic
 - [x] Docker Normal
 - [x] Windows

### CODE REVIEW

**Comment by NotExpectedYet** [comment]
Just needs double checking on windows but unix/docker all working now.

**Comment by davidzwa** [comment]
image
Continuous back and forth!

**Comment by NotExpectedYet** [comment]
> image
> Continuous back and forth!

Windows? weird because I left the default to pmem.

**Comment by davidzwa** [comment]
Ah sorry I was running from an old docker container, still memory doesnt work in nodemon version

**Comment by NotExpectedYet** [comment]
> Ah sorry I was running from an old docker container, still memory doesnt work in nodemon version.
> 
> Getting NaN's again in apex input for memory. Digging into it

Something windows side then as was all fine linux. Strange one like XD.

**Comment by davidzwa** [comment]
- [x] Tested on Docker for Windows,
- [x] Tested with nodemon Windows
- [ ] Test with pm2 windows
- [ ] Tested with WSL linux

File: client_src/js/settings.js [hunk 1]
```diff
@@ -426,50 +426,64 @@ const optionsCPU = {
...
+  }
+
+  const memoryInfo = systemInfo?.memoryInfo;
+  if (!!memoryInfo) {
+    const systemUsedRAM = memoryInfo.used;
+    const freeRAM = memoryInfo.free;
+
+    if (!!(currentProc?.memRss || currentProc?.mem)) {
+      let octoFarmRAM = currentProc?.memRss * 1000;
+      if (!currentProc.memRss || Number.isNaN(octoFarmRAM)) {
```
Comment (davidzwa) [inline client_src/js/settings.js:471]: Fallback depends on mem prop

File: client_src/js/settings.js [hunk 2]
```diff
@@ -426,50 +426,64 @@ const optionsCPU = {
...
-    const cpuLoad = systemInfo.cpuLoad.currentLoadSystem;
-    const octoLoad = systemInfo.sysProcess.cpuu;
-    const userLoad = systemInfo.cpuLoad.currentLoadUser;
-    const remain = cpuLoad + octoLoad + userLoad;
+  const currentProc = systemInfo?.currentProcess;
+  const cpuLoad = systemInfo?.cpuLoad;
+  if (!!cpuLoad?.currentLoadSystem && !!cpuLoad?.currentLoadUser) {
+    const systemLoad = cpuLoad.currentLoadSystem;
+    const userLoad = cpuLoad.currentLoadUser;
+    const octoLoad = !!currentProc?.cpuu ? currentProc.cpuu : 0;
```
Comment (davidzwa) [inline client_src/js/settings.js:457]: easier isNaN, null or undefined variant

File: client_src/js/settings.js [hunk 3]
```diff
@@ -426,50 +426,64 @@ const optionsCPU = {
...
+    procUptimeElem.innerHTML = Calc.generateTime(systemInfo.processUptime);
+  }
 
-    const cpuLoad = systemInfo.cpuLoad.currentLoadSystem;
-    const octoLoad = systemInfo.sysProcess.cpuu;
-    const userLoad = systemInfo.cpuLoad.currentLoadUser;
-    const remain = cpuLoad + octoLoad + userLoad;
+  const currentProc = systemInfo?.currentProcess;
+  const cpuLoad = systemInfo?.cpuLoad;
+  if (!!cpuLoad?.currentLoadSystem && !!cpuLoad?.currentLoadUser) {
```
Comment (davidzwa) [inline client_src/js/settings.js:454]: So we dont have to check for Object.keys().length > 0

File: client_src/js/settings.js [hunk 4]
```diff
@@ -426,50 +426,64 @@ const optionsCPU = {
...
+  optionsMemory,
 );
 systemChartMemory.render();
 setInterval(async function updateStatus() {
   let systemInfo = await Client.get("settings/sysInfo");
   systemInfo = await systemInfo.json();
 
-  if (
-    Object.keys(systemInfo).length === 0 &&
-    systemInfo.constructor === Object
```
Comment (davidzwa) [inline client_src/js/settings.js:443]: This made no sense to me, also given the fact that the if body was empty. Removed.

File: package-lock.json [hunk 5]
```diff
@@ -111,11 +112,7 @@
         "babel-external-helpers": "bin/babel-external-helpers.js"
       },
       "optionalDependencies": {
-        "@nicolo-ribaudo/chokidar-2": "2.1.8-no-fsevents",
```
Comment (davidzwa) [inline package-lock.json:114]: This dependency can cause npm install failures when babel is installed, so when installing non-production modules. That's why we need to be so careful with the lock version being 2 and to install only prod modules.

File: package-lock.json [hunk 6]
```diff
@@ -111,11 +112,7 @@
...
-        "chokidar": "^3.4.0"
```
Comment (davidzwa) [inline package-lock.json:115]: This dependency can cause npm install failures when babel is installed, so when installing non-production modules. That's why we need to be so careful with the lock version being 2 and to install only prod modules.

File: server_src/routes/settings.js [hunk 7]
```diff
@@ -12,24 +12,13 @@ const UserDB = require("../models/User.js");
...
-
-const systemInfo = require("../runners/systemInfo.js");
-
-const SystemInfo = systemInfo.SystemRunner;
-
-const settingsClean = require("../lib/dataFunctions/settingsClean.js");
-
-const { SettingsClean } = settingsClean;
-
+const { SystemRunner } = require("../runners/systemInfo.js");
```
Comment (davidzwa) [inline server_src/routes/settings.js:20]: I tend to simplify the includes and make them as compact as possible.
Also from other languages Ive learned to put node_modules at the top, followed by our own modules.

File: server_src/routes/settings.js [hunk 8]
```diff
@@ -280,22 +269,28 @@ router.post("/server/update", ensureAuthenticated, (req, res) => {
...
+
+  if (!!systemInformation) {
     sysInfo = {
       osInfo: systemInformation.osInfo,
       cpuInfo: systemInformation.cpuInfo,
       cpuLoad: systemInformation.cpuLoad,
       memoryInfo: systemInformation.memoryInfo,
       sysUptime: systemInformation.sysUptime,
-      sysProcess: systemInformation.sysProcess,
+      currentProcess: systemInformation.currentProcess,
```
Comment (davidzwa) [inline server_src/routes/settings.js:287]: Renamed

File: test/application/system-info-runner.test.js [hunk 9]
```diff
@@ -13,26 +13,44 @@ const diskFormats = [
...
+      } else {
+        return [systemUsedRAM, octoFarmRAM, freeRAM];
+      }
+    }
+    else {
+      return [systemUsedRAM, 0, freeRAM];
+    }
   } else {
-    return [systemUsedRAM, octoFarmRAM, freeRAM];
+    return [0, 0, 0];
```
Comment (davidzwa) [inline test/application/system-info-runner.test.js:53]: Desperation.

File: test/application/system-info-runner.test.js [hunk 10]
```diff
@@ -92,8 +110,8 @@ describe("SystemRunner", () => {
     expect(systemInfo.sysUptime.current).toBeGreaterThan(1617880660070); // ms time as of writing this test ^^
     expect(systemInfo.sysUptime.timezone).toBeTruthy();
     expect(systemInfo.processUptime).toEqual(expect.any(Number));
-    expect(systemInfo.sysProcess.pid).toBeGreaterThan(1);
-    expect(systemInfo.sysProcess.name).toEqual(expect.any(String));
+    expect(systemInfo.currentProcess.pid).toBeGreaterThan(1);
+    expect(systemInfo.currentProcess.name).toEqual(expect.any(String));
```
Comment (davidzwa) [inline test/application/system-info-runner.test.js:114]: Nice way to test JS code locally

**Review by davidzwa — COMMENTED** [review]
LGTM, but I committed so needs your approval as well

**Review by davidzwa — APPROVED** [review]

**Pull request merged**